### ISSUE TITLE
tf.reshape layer has problem in DNN module

### ISSUE THREAD

**Issue by shinyoungmin** [opencv/opencv#9992]
##### System information (version)

- OpenCV => 3.3.1
- Operating System / Platform => Ununtu 16.04
- Compiler => python 2.7 / gcc 5.4

##### Detailed description

I trained very simple standard CNN for MNIST dataset(or whatever) in Tensorflow, and saved the model.
All the post-saving processing like 'freeze graph' and 'optimize for inference' were done successfully.

In OpenCV (in c++ and python interface), importing model and forward() were run without error, but the resulting accuracy was very low.

So I checked all the trained weights and responses (output) of every layer in the network.
Everything was OK but when it comes to **reshape layer** right before the dense layer, the output of the reshape layer was  different from that of Tensorflow. 

I think the order that OpenCV DNN slices array differ from Tensorflow.

**Comment by dkurt** [comment]
@shinyoungmin, reshape layer doesn't modify the data nigher in OpenCV nor in TensorFlow. Show the list of TensorFlow nodes and OpenCV layers and names of tested layers that are different.

**Comment by shinyoungmin** [comment]
@dkurt, You're right. reshape layer doesn't modify data.
I wrote some toy example code to reproduce the difference.

# Tensorflow code
```
# imports
import numpy as np
import tensorflow as tf

# make input array to test
img = np.ones([1,5,5,3], dtype=np.float32)

# define placeholder
x = tf.placeholder(tf.float32, [None, 5, 5, 3])

# define net
conv = tf.layers.conv2d(inputs=x, filters=3, kernel_size=[3,3],
                        padding="same", name="test_conv")
vec = tf.reshape(conv, [-1, 5*5*3], name="test_reshape")

# run
sess = tf.Session()
sess.run(tf.global_variables_initializer())
result=sess.run(vec, feed_dict={x: img})
print(result)

# save graph
saver = tf.train.Saver()
saver.save(sess, 'tf_reshape_test.ckpt')
tf.train.write_graph(sess.graph_def, ".", "tf_reshape_test.pbtxt")
```

**After saving the graph, I did run freeze_graph and optimize_for_inference.**


# OpenCV code
```
import numpy as np
import cv2

# read saved Tensorflow network
net = cv2.dnn.readNetFromTensorflow("tf_reshape_test_optimized_graph.pb")

# input image
img = np.ones([5, 5, 3], dtype=np.float32)  # opencv image: height*width*nchannel

# convert to blob
blob = cv2.dnn.blobFromImage(img)   # blob.shape == 1x3x5x5
net.setInput(blob)

result = net.forward()
print result
```


**Finally the results are like this**

from tensorflow:
[[ 0.36306867  0.7826606   0.62073255  0.31695843  0.21590126 -0.00970408
   0.31695843  0.21590126 -0.00970408  0.31695843  0.21590126 -0.00970408
   0.36104068 -0.11288139 -0.42929643  0.1659295   1.24806952 -0.24212706
   0.17856239  0.72984922 -0.95878673  0.17856239  0.72984922 -0.95878673
   0.17856239  0.72984922 -0.95878673  0.49609366 -0.18229881 -0.76757675
   0.1659295   1.24806952 -0.24212706  0.17856239  0.72984922 -0.95878673
   0.17856239  0.72984922 -0.95878673  0.17856239  0.72984922 -0.95878673
   0.49609366 -0.18229881 -0.76757675  0.1659295   1.24806952 -0.24212706
   0.17856239  0.72984922 -0.95878673  0.17856239  0.72984922 -0.95878673
   0.17856239  0.72984922 -0.95878673  0.49609366 -0.18229881 -0.76757675
  -0.49562851  1.36638308 -0.63177717 -0.20839173  1.20973837 -1.00589108
  -0.20839173  1.20973837 -1.00589108 -0.20839173  1.20973837 -1.00589108
   0.60907102  0.10145521 -0.40738222]]


fron opencv dnn:
[[ 0.3630687   0.31695843  0.31695843  0.31695843  0.36104068  0.1659295
   0.17856243  0.17856243  0.17856243  0.49609372  0.1659295   0.17856243
   0.17856243  0.17856243  0.49609372  0.1659295   0.17856243  0.17856243
   0.17856243  0.49609372 -0.49562848 -0.20839173 -0.20839173 -0.20839173
   0.60907102  0.78266072  0.21590126  0.21590126  0.21590126 -0.11288136
   1.24806952  0.72984922  0.72984922  0.72984922 -0.18229881  1.24806952
   0.72984922  0.72984922  0.72984922 -0.18229881  1.24806952  0.72984922
   0.72984922  0.72984922 -0.18229881  1.36638308  1.20973837  1.20973837
   1.20973837  0.10145521  0.62073267 -0.0097041  -0.0097041  -0.0097041
  -0.42929643 -0.24212706 -0.95878673 -0.95878673 -0.95878673 -0.76757681
  -0.24212706 -0.95878673 -0.95878673 -0.95878673 -0.76757681 -0.24212706
  -0.95878673 -0.95878673 -0.95878673 -0.76757681 -0.63177717 -1.00589097
  -1.00589097 -1.00589097 -0.40738222]]


As we can see, 2 results share the same values (considering numerical precision). But the numbers appear in different position. They are actually 1-d array, so they should be matched in the same position.

**Comment by dkurt** [comment]
@shinyoungmin, try to add [tf.transpose](https://www.tensorflow.org/versions/master/api_docs/python/tf/transpose) node as `tf.transpose(conv, [0, 1, 2, 3])` to indicate that you want to use data in NHWC layout.

**Comment by shinyoungmin** [comment]
@dkurt , if I add tr.transpose node after convolution and save the modified graph in Tensorflow, the loaded graph in OpenCV will also be changed; so the outputs will be different again :)

The point here is not using the reshaped array as final output. Generally the tf.reshape layer is used to flatten the tensor and feed it to the fully-connected(dense) layer afterwards. In such a case, in OpenCV, the fully-connected layer will be provided with differently reshaped arrays as inputs.

So, to get the same result with Tensorflow, transposing tensor should be done internally in DNN module when the graph has reshape ops.

**Comment by shinyoungmin** [comment]
@GPhilo
To cut a long story short, I gave up using reshape + dense layer since they are problematic in OpenCV. Instead I recommend using fully-convolutional in the last layers.

I am sure your wrong result starts from the reshape layer.
Manually inserting a conversion of NHWC to NCHW in the middle of network is contradiction.

**Comment by GPhilo** [comment]
So, it turns out we were the cause of our own problems. Our console output was transposing a matrix to display it and that was breaking the computation. Taking that out, now the dense layers work as expected.

**Comment by dkurt** [comment]
@shinyoungmin, May I ask you to look at this example: https://github.com/opencv/opencv/issues/10065#issuecomment-343427950.

**Comment by shinyoungmin** [comment]
@dkurt 
Wow, your suggestion to insert transpose in Tensorflow works well. Now I have 2 nice workarounds which can connect conv layers to the final outputs. Thanks.!

I still think this issue should be solved in the next release for the consistency with {BGR to RGB} and {NCHW to NHWC} in the input which are internally taken care of by OpenCV.

**Comment by czw90130** [comment]
@dkurt 
I tested your example in #10065 and got Segmentation fault in OpenCV version. I think the problem should in tf.transpose function.

@shinyoungmin 
use  a conv2d with kernel_size = [5, 5]. should slove the problem.
## TF CODE:
```python
# imports
import numpy as np
import tensorflow as tf

# make input array to test
img = np.ones([1,5,5,3], dtype=np.float32)

# define placeholder
x = tf.placeholder(tf.float32, [None, 5, 5, 3], name="input")

# define net
conv = tf.layers.conv2d(inputs=x, filters=3, kernel_size=[3,3],
                        padding="SAME", name="test_conv")

# vec = tf.contrib.layers.flatten(conv)
# fc = tf.contrib.layers.fully_connected(vec,8)
# fc = tf.identity(fc, name="output")
# conv_nchw = tf.transpose(conv, [0,3,1,2])
# vec = tf.reshape(conv, [-1, 5*5*3], name="output")

conv_vec = tf.layers.conv2d(inputs=conv, filters=5*5*3, kernel_size=[conv.shape[1],conv.shape[2]],
                        padding="VALID", name="conv_reshape")
vec = tf.reshape(conv_vec, [-1, 5*5*3], name="output")

# run
sess = tf.Session()
sess.run(tf.global_variables_initializer())
result=sess.run(vec, feed_dict={x: img})
print(result)

# save graph
saver = tf.train.Saver()
saver.save(sess, './model.ckpt')
tf.train.write_graph(sess.graph_def, ".", "graph.pbtxt")
```

**Comment by dkurt** [comment]
@shinyoungmin, @GPhilo, @czw90130, may I ask you to test the following PR https://github.com/opencv/opencv/pull/10364 for your test cases? Please let us know if it can solve it or not. Thanks!
